cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java

```diff
@@ -0,0 +1,381 @@
+  private final Transaction transaction;
+  private final StreamAdmin streamAdmin;
+  private final DatasetInstantiator datasetInstantiator;
+  private boolean stopped;
+  private ApacheSparkContext apacheSparkContext;
+
+  public ExecutionSparkContext(SparkSpecification specification, Id.Program programId, RunId runId,
+                               ClassLoader programClassLoader, long logicalStartTime,
+                               Map<String, String> runtimeArguments,
+                               Transaction transaction, DatasetFramework datasetFramework,
+                               DiscoveryServiceClient discoveryServiceClient,
+                               MetricsCollectionService metricsCollectionService,
+                               Configuration hConf, StreamAdmin streamAdmin) {
+    this(specification, programId, runId, programClassLoader, logicalStartTime, runtimeArguments,
+         transaction, datasetFramework, discoveryServiceClient,
+         createMetricsContext(metricsCollectionService, programId, runId),
+         createLoggingContext(programId, runId), hConf, streamAdmin);
+  }
+
+  public ExecutionSparkContext(SparkSpecification specification, Id.Program programId, RunId runId,
+                                ClassLoader programClassLoader, long logicalStartTime,
+                                Map<String, String> runtimeArguments,
+                                Transaction transaction, DatasetFramework datasetFramework,
+                                DiscoveryServiceClient discoveryServiceClient, MetricsContext metricsContext,
+                                LoggingContext loggingContext, Configuration hConf, StreamAdmin streamAdmin) {
+    super(specification, programId, runId, programClassLoader, logicalStartTime,
+          runtimeArguments, discoveryServiceClient, metricsContext, loggingContext);
+
+    this.datasets = new HashMap<>();
+    this.hConf = hConf;
+    this.transaction = transaction;
+    this.streamAdmin = streamAdmin;
+    this.datasetInstantiator = new DatasetInstantiator(programId.getNamespace(), datasetFramework,
+                                                       programClassLoader, getOwners(), getMetricsContext());
+  }
+
+  @Override
+  public <T> T readFromDataset(String datasetName, Class<?> kClass, Class<?> vClass) {
+    Map<String, String> datasetArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
+
+    // Verify the dataset exists. If it doesn't exist, exception will be thrown
+    Closeables.closeQuietly(createDataset(datasetName, datasetArgs));
+
+    // Clone the configuration since it's dataset specification and shouldn't affect the global hConf
+    Configuration configuration = new Configuration(hConf);
+    SparkDatasetInputFormat.setDataset(configuration, datasetName, datasetArgs);
+    return getApacheSparkContext().createRDD(SparkDatasetInputFormat.class, kClass, vClass, configuration);
```

Comment:
The way for registering usage of dataset vs stream is not consistent. For dataset, the DS service do it on the server side, hence no need to do registration here.